torch/optim/nadam.py

```diff
@@ -116,23 +126,155 @@ def step(self, closure=None):
+                  momentum_decay=group['momentum_decay'],
+                  eps=group['eps'],
+                  foreach=group['foreach'])
 
             # update mu_product
-            for p, mu_product in zip(params_with_grad, mu_products):
+            for p in params_with_grad:
                 state = self.state[p]
                 state['mu_product'] = state['mu_product'] * beta1 * \
                     (1. - 0.5 * (0.96 ** (state['step'] * group['momentum_decay'])))
 
         return loss
+
+
+def nadam(params: List[Tensor],
+          grads: List[Tensor],
+          exp_avgs: List[Tensor],
+          exp_avg_sqs: List[Tensor],
+          mu_products: List[float],
+          state_steps: List[int],
+          foreach: bool = None,
+          *,
+          beta1: float,
+          beta2: float,
+          lr: float,
+          weight_decay: float,
+          momentum_decay: float,
+          eps: float):
+    r"""Functional API that performs NAdam algorithm computation.
+
+    See :class:`~torch.optim.NAdam` for details.
+    """
+
+    if foreach is None:
+        # Placeholder for more complex foreach logic to be added when value is not set
+        foreach = False
+
+    if foreach and torch.jit.is_scripting():
+        raise RuntimeError('torch.jit.script not supported with foreach optimizers')
+
+    if foreach and not torch.jit.is_scripting():
+        func = _multi_tensor_nadam
+    else:
+        func = _single_tensor_nadam
+
+    func(params,
+         grads,
+         exp_avgs,
+         exp_avg_sqs,
+         mu_products,
+         state_steps,
+         beta1=beta1,
+         beta2=beta2,
+         lr=lr,
+         weight_decay=weight_decay,
+         momentum_decay=momentum_decay,
+         eps=eps)
+
+
+def _single_tensor_nadam(params: List[Tensor],
+                         grads: List[Tensor],
+                         exp_avgs: List[Tensor],
+                         exp_avg_sqs: List[Tensor],
+                         mu_products: List[float],
+                         state_steps: List[int],
+                         *,
+                         beta1: float,
+                         beta2: float,
+                         lr: float,
+                         weight_decay: float,
+                         momentum_decay: float,
+                         eps: float):
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        exp_avg = exp_avgs[i]
+        exp_avg_sq = exp_avg_sqs[i]
+        mu_product = mu_products[i]
+        step = state_steps[i]
+
+        bias_correction2 = 1 - beta2 ** step
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # calculate the momentum cache \mu^{t} and \mu^{t+1}
+        mu = beta1 * (1. - 0.5 * (0.96 ** (step * momentum_decay)))
+        mu_next = beta1 * (1. - 0.5 * (0.96 ** ((step + 1) * momentum_decay)))
+        mu_product = mu_product * mu
+        mu_product_next = mu_product * mu * mu_next
```

Comment:
I think this step is actually a bug - `mu_product_next` should just be `mu_product * mu_next`, so the additional step actually did not have to be added on line 259 for parity, should I file an issue for this?

Reply:
Yes please file an issue! Thanks for looking into the computation in details.